### PR TITLE
Added ROUTE::OPTIONAL_SECURED

### PR DESCRIPTION
**Pull request by vjirovsky**
This option for route allows to access route via both HTTP&HTTPS without
redirection.
Href URLs will be generated in currently used protocol.

### CODE REVIEW

File: Nette/Application/Routers/Route.php [hunk 1]
```diff
@@ -251,13 +253,15 @@ public function match(Nette\Http\IRequest $httpRequest)
 			unset($params[self::PRESENTER_KEY]);
 		}
 
+		$this->httpRequestSecured = $httpRequest->isSecured();
```
Comment (fprochazka) [inline Nette/Application/Routers/Route.php]: Please do not save the state here, it's not neccessary and makes it harder to test router.

**Comment by fprochazka** [comment]
@vjirovsky great idea :+1:

**Comment by f3l1x** [comment]
I use same code as @Majkl578 

```php
if ($request->isSecured()) {
    Route::$defaultFlags |= Route::SECURED;
}
```

**Comment by hrach** [comment]
This issue is mostly problematic because developers realize it quite late - after deploying application into a production. Till then, they expect routing will work **as same as it ignores domain names**. The route doesn't force any protocol, so this behavior should be the default one.

**Comment by f3l1x** [comment]
@hrach I force it at Nginx mainly.

**Comment by hrach** [comment]
@f3l1x please follow the text, so do I, so does everyone.

**Comment by JanTvrdik** [comment]
> The route doesn't force any protocol, so this behavior should be the default one.

:+1: Can anyone think of some negative impact?

Current default behavior (correct me please if I'm wrong) is to accept (*match*) both HTTP and HTTPS requests but generate always HTTP unless SECURED flag is set.

Behavior matrix to summarize

| Accept         | Generates | Note                       |
|----------------|-----------|----------------------------| 
| HTTP only      | HTTP      | now impossible             |
| HTTP and HTTPS | HTTP      | current default behavior   |
| HTTP and HTTPS | HTTPS     | behavior with SECURED flag |
| HTTP and HTTPS | COPY      | proposed behavior          |
| HTTPS only     | HTTPS     | now impossible             |

**Comment by Majkl578** [comment]
I'm not sure whether running both HTTP & HTTPS is still an issue nowadays... Web is moving towards TLS only -- we have HTTP/2 that requires TLS, we also have free certificates provided by Let's Encrypt.

**Comment by milo** [comment]
Idea: What about not to update Route, but:
a) add example code to sandbox's bootstrap commented out, or
b) adjust router factory in sandbox to control this functionality

It will warn people that there is something to configure.

**Comment by dg** [comment]
IMHO default behavior can be changed to *copy* protocol when route mask doesn't contain domain or when generated domain matches current domain.

**Comment by JanTvrdik** [comment]
> IMHO default behavior can be changed to copy protocol 

Can this be done in Nette 2.4?

**Pull request closed without merging**